client/src/main/java/org/eclipse/hono/client/impl/HonoClientImpl.java

```diff
@@ -916,19 +917,52 @@ public final void shutdown() {
      */
     @Override
     public final void shutdown(final Handler<AsyncResult<Void>> completionHandler) {
-
         if (shuttingDown.compareAndSet(Boolean.FALSE, Boolean.TRUE)) {
-            context.runOnContext(shutDown -> {
-                if (isConnectedInternal()) {
-                    shutdownConnection(completionHandler);
-                } else {
-                    LOG.info("connection to server [{}:{}] already closed", connectionFactory.getHost(),
-                            connectionFactory.getPort());
-                    completionHandler.handle(Future.succeededFuture());
-                }
+            context.runOnContext(shutDownResult -> {
+                closeConnection(completionHandler);
+            });
+        } else {
+            completionHandler.handle(Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_CONFLICT, "already in the middle of a shutdown operation")));
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect() {
+        disconnect(disconnectResult -> {
+            if (disconnectResult.succeeded()) {
+                LOG.info("successfully disconnected from the server");
+            } else {
+                LOG.error("could not disconnect from the server", disconnectResult.cause());
+            }
+        });
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public final void disconnect(final Handler<AsyncResult<Void>> completionHandler) {
+        if (disconnecting.compareAndSet(Boolean.FALSE, Boolean.TRUE)) {
+            context.runOnContext(disconnectResult -> {
+                disconnecting.compareAndSet(Boolean.TRUE, Boolean.FALSE);
+                closeConnection(completionHandler);
             });
         } else {
-            completionHandler.handle(Future.failedFuture(new IllegalStateException("already shutting down")));
+            completionHandler.handle(Future.failedFuture(new ClientErrorException(1, "already in the middle of a disconnect operation")));
+        }
+    }
+
+    //-----------------------------------< private methods >---
+
+    private void closeConnection(final Handler<AsyncResult<Void>> completionHandler) {
+        if (isConnectedInternal()) {
+            shutdownConnection(completionHandler);
```

Comment:
I think the `shutdownConnection` code can be inlined here (and `shutdownConnection()`removed) to avoid naming confusion concerning these methods.